protocols/mdns/src/behaviour.rs

```diff
@@ -115,29 +126,54 @@ pub struct Mdns {
 
     /// Discovery timer.
     timeout: Timer,
+
+    /// IP Type.
+    domain: IPType,
 }
 
 impl Mdns {
     /// Builds a new `Mdns` behaviour.
     pub async fn new(config: MdnsConfig) -> io::Result<Self> {
-        let recv_socket = {
-            let socket = Socket::new(
-                Domain::ipv4(),
-                Type::dgram(),
-                Some(socket2::Protocol::udp()),
-            )?;
-            socket.set_reuse_address(true)?;
-            #[cfg(unix)]
-            socket.set_reuse_port(true)?;
-            socket.bind(&SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 5353).into())?;
-            let socket = socket.into_udp_socket();
-            socket.set_multicast_loop_v4(true)?;
-            socket.set_multicast_ttl_v4(255)?;
-            Async::new(socket)?
+        let recv_socket = match config.domain {
+            IPType::IPv4 => {
+                let socket = Socket::new(
+                    Domain::ipv4(),
+                    Type::dgram(),
+                    Some(socket2::Protocol::udp()),
+                )?;
+                socket.set_reuse_address(true)?;
+                #[cfg(unix)]
+                socket.set_reuse_port(true)?;
+                socket.bind(&SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 5353).into())?;
+                let socket = socket.into_udp_socket();
+                socket.set_multicast_loop_v4(true)?;
+                socket.set_multicast_ttl_v4(255)?;
+                Async::new(socket)?
+            }
+            IPType::IPv6 => {
+                let socket = Socket::new(
+                    Domain::ipv6(),
+                    Type::dgram(),
+                    Some(socket2::Protocol::udp()),
+                )?;
+                socket.set_reuse_address(true)?;
+                #[cfg(unix)]
+                socket.set_reuse_port(true)?;
+                socket.bind(&SocketAddr::new(Ipv6Addr::UNSPECIFIED.into(), 5353).into())?;
+                let socket = socket.into_udp_socket();
+                socket.set_multicast_loop_v6(true)?;
+                Async::new(socket)?
+            }
         };
-        let send_socket = {
-            let socket = UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0))?;
-            Async::new(socket)?
+        let send_socket = match config.domain {
+            IPType::IPv4 => {
+                let socket = std::net::UdpSocket::bind((Ipv4Addr::UNSPECIFIED, 0))?;
+                Async::new(socket)?
+            }
+            IPType::IPv6 => {
+                let socket = std::net::UdpSocket::bind((Ipv6Addr::UNSPECIFIED, 0))?;
+                Async::new(socket)?
```

Comment:
same here, binding to `Ipv6Addr::UNSPECIFIED` should allow sending ipv4 packets. please test that this is indeed the case

Reply:
It works on Linux, macOS has problem with joining multicast on IPv6, while switching to IPv6Addr on macOS I got os error 22. Should this be compatible with macOS?
If not I will go with your solution Ippv6Addr::UNSPECIFIED, but if it should i propose:
let send_socket = {
            let addrs = [
                SocketAddr::new(IpAddr::V4(Ipv4Addr::UNSPECIFIED), 0),
                SocketAddr::new(IpAddr::V6(Ipv6Addr::UNSPECIFIED), 0),
            ];

            let socket = std::net::UdpSocket::bind(&addrs[..])?;
            Async::new(socket)?
        };